src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala

```diff
@@ -843,6 +843,17 @@ abstract class BCodeHelpers extends BCodeIdiomatic with BytecodeWriters {
       }
     }
 
+    /*
+     * must-single-thread
+     */
+    def emitParamNames(jmethod: asm.MethodVisitor, params: List[ValDef]) = {
+      for (param <- params) {
+        var access = asm.Opcodes.ACC_FINAL
+        if (param.mods.isSynthetic)
+          access |= asm.Opcodes.ACC_SYNTHETIC
```

Comment:
echoing my earlier comment:

This is a known area of confusion in scalac: we have two flags for symbols `SYNTHETIC` and `ARTIFACT`. The distinction is that only things marked `ARTIFACT` should be marked `ACC_SYNTHETIC` in the bytecode, see https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/api/FlagSets.scala#L229-L248 and https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala#L310.

It is neither well-specified nor well-tested what definitions in a classfile exactly get the `ACC_SYNTHETIC`, but in order to retain at least the current consistency we should only add the flag here for params that are `isArtifact`.

Also, another non-obvious thing: instead of looking at the flags in `mods`, you should look at the symbol (`param.symbol.isArtifact`); the modifiers are generally used until Typer, after that the symbol should be the main source of information.

cc @retronym – please correct me if I'm wrong!

Thinking a bit more, using `ARTIFACT` would mean that a lot of synthetic params are not marked `ACC_SYNTHETIC`: `newSyntheticValueParams` marks the symbols `SYNTHETIC` only, it's used for example for default getters and delambdafy methods.

Given that we already don't emit `ACC_SYNTHETIC` for a lot of scala-synthetic definitions (default getter methods for example), I'd say that's fine, we're not making the situation fundamentally worse.